nixos/home-assistant: do not set server_host by default

71rd opened the pull request:
The option `services.home-assistant.config.http.server_host' has been
deprecated by home-assistant in version 2025.12  and will be removed with
version 2026.06. 
See commit:
https://github.com/home-assistant/core/commit/07c4c58ce45cd54e067ab01e7fcc6e163cdee0c7

This commit removes the default value from the config and introduces a
warning if set manually. It also addss an assert that prevents settings
the option starting with HA version 2026.06

This option can not be marked as removed with lib.mkRemovedOptionModule
as `services.home-assistant.config` has the type yaml and is not a submodule

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [x] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dotlambda reviewed (changes requested):
I'm fine with removing the default so that people who don't set `server_host` don't see a warning in the web interface.
But I don't think the warnings make sense because I'm not sure that `server_host` will actually be removed upstream. See the discussion in the Matrix room.

mweinelt commented:
Even if it goes away I think it might be simple enough to patch it back in if we want it. I'm not sure why they'd remove it from core, it shouldn't break anything. That might just be that they don't support core and therefore don't care.

71rd commented:
> I totally agree. How does formats.yaml handle null values? If it simply doesn't set those keys, then we can just change the default to null.

Thats a great idea, setting the value to null appears to work as you suggested.

Im slightly in favour of stopping to advertise the option wholly since it appears to be at least unwanted since the deprecation has not be reversed [yet](https://github.com/home-assistant/core/issues/157961), but am also happy with this solution for now

dotlambda commented:
Why not set the default to `null`? That's a non-breaking change since the default is already to bind to all interfaces and it removes the warning for people who don't set the option.
But yeah, I don't think any warnings or release notes should be added until https://github.com/home-assistant/core/issues/157961 is resolved.

71rd committed:
nixos/home-assistant: do not set server_host by default

Allowing `services.home-assistant.config.http.server_host` to be set
to `null` and making this default.

Setting the value to null leads to it not being set in the by formats.ya
This should not change the default behaviour of home-assistant, as
by default the process binds to all available addresses.

This change avoids creating a warning in home-assistant about the deprecation
of the server_host option and expected removal in 2026.06

An open issue for upstream around the removal can be found here:
https://github.com/home-assistant/core/issues/157961
